File: InterviewCake/Python/BinaryTree/Superbalanced_Binary_Tree.py
#This question is focussed on the superbalance
#property of a binary tree.

'''
Reference: https://www.interviewcake.com/question/python3/balanced-binary-tree?utm_source=weekly_email&utm_source=drip&utm_campaign=weekly_email&utm_campaign=Interview%20Cake%20Weekly%20Problem%20%23329:%20Balanced%20Binary%20Tree&utm_medium=email&utm_medium=email&__s=ap5rqqpkspoovezbmxvs

Definition of the problem: A tree is 'superbalanced' if the difference
between the depths of any 2 leaf nodes is no greater than 1.


I go to all the leaf nodes, each time I call the function:
I update the depth by 1

Example, consider the following tree:

        1
      /    \
      2      3
    /  \    / \
    4   5  6   7
  /  \
  8   9


We are going to build the tree iteratively in this case
'''
class Node:
    def __init__(self,node):
        self.left = None
        self.right = None
        self.node = node

#Recursive solution
def max_depth(tree):
    '''
    1. Query on and on until graph does not have children anymore.
       each time add +1
       when you stop that value gives you the depth
    '''
    minDepth = float("Inf")
    maxDepth = 0

    def findDepth(node,depth,minDepth,maxDepth):
        #if the condition is violated we return
        if (maxDepth-minDepth)>1:
            return

        if node.left:
            findDepth(node.left, depth+1, minDepth, maxDepth)
        if node.right:
            findDepth(node.right, depth+1, minDepth, maxDepth)

        #At the leaf node, the current node does not have
        if not node.right and not node.left:
            #go down the tree.
            if depth>maxDepth:
                maxDepth=depth
            if depth<minDepth:
                minDepth=depth

    findDepth(tree,0,minDepth,maxDepth)
    print ('minDepth',minDepth)
    print ('maxDepth',maxDepth)
    return maxDepth - minDepth <=1


if __name__ == "__main__":
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.right = Node(5)
    root.right.left = Node(6)
    root.right.right = Node(7)
    root.left.left.left = Node(8)
    root.left.left.right = Node(9)


    #Call the function here:
    print ('The tree is superbalance:',max_depth(root))

File: InterviewCake/Python/BinaryTree/Superbalanced_Binary_Tree.cpp
//Rewriting the code in .cpp
//Rewriting the code: of Superbalanced.py, Superbalanced.cpp

#include <iostream>
#include <vector>
#include <limits>

using namespace std;

struct Node
{
public:
  int value;
  Node* left;
  Node* right;

  Node(int value)
  {
    this->value = value;
    left = NULL;
    right = NULL;
  }
};


/*
Node* buildNode(int value)
{
  Node* node = new Node();
  node->value = value;
  node->left = NULL;
  node->right = NULL;
};
*/

bool max_Depth(Node* node)
{
  //Query on and on until graph does not have any children.
  //each time add +1 Recursive call
  //when you stop that value gives you the depth
  double minDepth = numeric_limits<double>::max();
  double maxDepth = 0;

  void findMinMaxDepth(Node* tree, int depth,double maxDepth, double minDepth)
  {
    if (maxDepth-minDepth >1)
    {
      return;
    }
    if (node->left)
    {
      findMinMaxDepth(node->left,depth+1);
    }
    if (node->right)
    {
      findMinMaxDepth(node->right,depth+1);
    }
    if (!node->left && !node->right)
    {
      if (depth>maxDepth)
      {
        madDepth = depth;
      }
      if (depth<minDepth)
      {
        minDepth=depth;
      }
    }
  }
  findMinMaxDepth(&tree,0);
  cout << "Max Depth:" << maxDepth << endl;
  cout << "Min Depth:" << minDepth << endl;
  return (maxDepth-minDepth) ? true:false;
};

int main()
{
  struct Node* root = new Node(1);
  root->left = new Node(2);
  root->right = new Node(3);
  root->left->left = new Node(4);
  root->left->right = new Node(5);
  root->right->left = new Node(6);
  root->right->right = new Node(7);
  root->left->left->left = new Node(8);
  root->left->left->right = new Node(9);
  cout <<"The tree is superbalanced:" <<max_depth(root)<<endl;
  return 0;
}
